// github-fetcher-mcp/dist/tools/fetch-subdir-tree.js
import { z } from 'zod';
import { formatTree } from '../services/tree-formatter.js';
/**
 * Creates the fetch-subdir-tree tool
 * @param fetcher - GitHub fetcher service
 * @param repoIdentifier - Optional repository identifier from CLI args
 */
export function createFetchSubdirTreeTool(fetcher, repoIdentifier) {
    // Define schema based on whether repoIdentifier is provided
    const inputSchema = repoIdentifier
        ? z.object({
            dirPath: z.string().describe('Path to the directory in the repository (empty for root)'),
        })
        : z.object({
            ownerName: z.string().describe('Repository owner name'),
            repoName: z.string().describe('Repository name'),
            branchName: z.string().optional().default('main').describe('Branch name (default: main)'),
            dirPath: z.string().describe('Path to the directory in the repository (empty for root)'),
        });
    return {
        name: 'fetch-subdir-tree',
        description: repoIdentifier
            ? `Fetches directory tree from GitHub repository ${repoIdentifier.owner}/${repoIdentifier.repo}/${repoIdentifier.branch}`
            : 'Fetches directory tree from a GitHub repository',
        inputSchema,
        async handler(input) {
            const owner = repoIdentifier?.owner ?? input.ownerName;
            const repo = repoIdentifier?.repo ?? input.repoName;
            const branch = repoIdentifier?.branch ?? (input.branchName || 'main');
            const dirPath = input.dirPath;
            const treeResponse = await fetcher.fetchDirectoryTree(owner, repo, branch, dirPath);
            const formattedTree = formatTree(treeResponse, dirPath);
            return {
                content: [
                    {
                        type: 'text',
                        text: formattedTree,
                    },
                ],
            };
        },
    };
}
